src/LibFTD2XX.jl

```diff
@@ -1,260 +1,663 @@
-  end
-  handle
+Open a [`D2XXDevice`](@ref) for reading and writing using [`FT_OpenEx`](@ref).
+Cannot be used to open the same device twice.
+
+See also: [`isopen`](@ref), [`close`](@ref)
+"""
+function Base.open(d::D2XXDevice)
+  isopen(d) && throw(D2XXException("Device already open."))
+  fthandle(d, FT_Open(deviceidx(d)))
+  return
 end
 
+"""
+    open(str::AbstractString, openby::FTOpenBy) -> FT_HANDLE
+
+Create an open [`FT_HANDLE`](@ref) for reading and writing using 
+[`FT_OpenEx`](@ref). Cannot be used to open the same device twice.
+
+# Arguments
+ - `str::AbstractString` : Device identifier. Type depends on `openby`
+ - `openby::FTOpenBy` : Indicator of device identifier `str` type.
+
+See also: [`isopen`](@ref), [`close`](@ref)
+"""
+Base.open(str::AbstractString, openby::FTOpenBy) =  FT_OpenEx(str, DWORD(openby))
+
+
+"""
+    close(d::D2XXDevice)
+
+Close a [`D2XXDevice`](@ref) using [`FT_Close`](@ref). Does not perform a 
+[`flush`](@ref) first.
+"""
+Base.close(d::D2XXDevice) = close(fthandle(d))
+
+"""
+    close(handle::FT_HANDLE)
+
+Close an [`FT_HANDLE`](@ref) using [`FT_Close`](@ref). Does not perform a 
+[`flush`](@ref) first.
+"""
 function Base.close(handle::FT_HANDLE)
-  status = ccall(cfunc[:FT_Close], cdecl, FT_STATUS, (FT_HANDLE, ),
-                                                       handle)
-  FT_STATUS_ENUM(status) == FT_OK || throw(FT_STATUS_ENUM(status))
-  handle.p = C_NULL
+  if isopen(handle)
+    FT_Close(handle)
+  end
   return
 end
 
+"""
+    bytesavailable(d::D2XXDevice)
+
+See also: [`D2XXDevice`](@ref), [`isopen`](@ref), [`open`](@ref), 
+[`readavailable`](@ref), [`read`](@ref)
+"""
+Base.bytesavailable(d::D2XXDevice) = bytesavailable(fthandle(d))
+
+"""
+    bytesavailable(handle::FT_HANDLE)
+
+See also: [`FT_HANDLE`](@ref), [`isopen`](@ref), [`open`](@ref), 
+[`readavailable`](@ref), [`read`](@ref)
+"""
+function Base.bytesavailable(handle::FT_HANDLE)
+  isopen(handle) || throw(D2XXException("Device must be open to check bytes available."))
+  FT_GetQueueStatus(handle)
+end
+
+"""
+    eof(d::D2XXDevice) -> Bool
+
+Indicates if any bytes are available to be read from an open 
+[`D2XXDevice`](@ref). Non-blocking.
+
+See also: [`isopen`](@ref), [`open`](@ref), 
+[`readavailable`](@ref), [`read`](@ref)
+"""
+Base.eof(d::D2XXDevice) = eof(fthandle(d))
+
+"""
+    eof(d::D2XXDevice) -> Bool
+
+Indicates if any bytes are available to be read from an open 
+[`FT_HANDLE`](@ref). Non-blocking.
+
+See also: [`isopen`](@ref), [`open`](@ref), 
+[`readavailable`](@ref), [`read`](@ref)
+"""
+Base.eof(handle::FT_HANDLE) = (bytesavailable(handle) == 0)
+
+
+"""
+    readbytes!(d::D2XXDevice, b::AbstractVector{UInt8}, nb=length(b))
+
+See also: [`D2XXDevice`](@ref).
+"""
+Base.readbytes!(d::D2XXDevice, b::AbstractVector{UInt8}, nb=length(b)) =
+readbytes!(fthandle(d), b, nb)
+
+"""
+    readbytes!(handle::FT_HANDLE, b::AbstractVector{UInt8}, nb=length(b))
+
+See also: [`FT_HANDLE`](@ref)
```

Comment:
Might be worth commenting that the vector will be resized if not large enough

Reply:
Also, if you resize an array I assume that it might change memory address. It will still be correctly bound (and thus still mutating the same object) right?

Reply:
> Might be worth commenting that the vector will be resized if not large enough

This is the standard behaviour for `readbytes!` for an `<: IO` and is already documented. From the standard repl:
```julia
julia> using LibFTD2XX
[ Info: Recompiling stale cache file C:\Users\Reuben\.julia\compiled\v1.0\LibFTD2XX\U18mP.ji for LibFTD2XX [505cea47-0b8c-5d8e-a4f7-f3ac0095ebab]

help?> readbytes!
search: readbytes!

  readbytes!(stream::IO, b::AbstractVector{UInt8}, nb=length(b))

  Read at most nb bytes from stream into b, returning the number of bytes
  read. The size of b will be increased if needed (i.e. if nb is greater than
  length(b) and enough bytes could be read), but it will never be decreased.

  ────────────────────────────────────────────────────────────────────────────

  readbytes!(stream::IOStream, b::AbstractVector{UInt8}, nb=length(b); all::Bool=true)

  Read at most nb bytes from stream into b, returning the number of bytes
  read. The size of b will be increased if needed (i.e. if nb is greater than
  length(b) and enough bytes could be read), but it will never be decreased.

  See read for a description of the all option.

  ────────────────────────────────────────────────────────────────────────────

  readbytes!(d::D2XXDevice, b::AbstractVector{UInt8}, nb=length(b))

  See also: D2XXDevice.

  ────────────────────────────────────────────────────────────────────────────

  readbytes!(handle::FT_HANDLE, b::AbstractVector{UInt8}, nb=length(b))

  See also: FT_HANDLE
```
Is this sufficiently non-ambiguous? Seems pointless to add an identical comment to this method


> Also, if you resize an array I assume that it might change memory address. It will still be correctly bound (and thus still mutating the same object) right?

 I don't follow?

Reply:
I've updated the `readbytes!` comment in d61c7d01d9241bebd9544ac580943855fe9de38d